SmartDeviceLink/SDLUploadFileOperation.m

```diff
@@ -64,99 +69,165 @@ - (void)sdl_sendPutFiles:(NSArray<SDLPutFile *> *)putFiles withCompletion:(SDLFi
-                                   }
-
-                                   dispatch_group_leave(putFileGroup);
-                               }];
+        [self.connectionManager sendManagerRequest:putFile withResponseHandler:^(__kindof SDLRPCRequest *_Nullable request, __kindof SDLRPCResponse *_Nullable response, NSError *_Nullable error) {
+            typeof(weakself) strongself = weakself;
+
+            // Check if the upload process has been cancelled by another packet. If so, stop the upload process.
+            if (strongself.isCancelled) {
+                // TODO: Is this the right way to handle this case? Should we just abort everything in the future? Should we be deleting what we sent? Should we have an automatic retry strategy based on what the error was?
+                stop = YES;
+            }
+
+            // If the SDL Core returned an error, cancel the upload the process in the future
+            if (error != nil || response == nil || ![response.success boolValue]) {
+                stop = YES;
+                streamError = error;
+                dispatch_group_leave(putFileGroup);
+                BLOCK_RETURN;
+            }
+
+            if (stop) {
+                dispatch_group_leave(putFileGroup);
+                BLOCK_RETURN;
+            }
+
+            // If no errors, watch for a response containing the amount of storage left on the SDL Core
+            if ([self sdl_newHighestCorrelationID:request highestCorrelationIDReceived:highestCorrelationIDReceived]) {
+                highestCorrelationIDReceived = [request.correlationID integerValue];
+                bytesAvailable = [(SDLPutFileResponse *)response spaceAvailable].unsignedIntegerValue;
+            }
+
+            dispatch_group_leave(putFileGroup);
+        }];
     }
-
     dispatch_group_leave(putFileGroup);
 }
 
-+ (NSArray<SDLPutFile *> *)sdl_splitFile:(SDLFile *)file mtuSize:(NSUInteger)mtuSize {
-    NSData *fileData = [file.data copy];
-    NSUInteger currentOffset = 0;
-    NSMutableArray<SDLPutFile *> *putFiles = [NSMutableArray array];
+/**
+ Opens a socket for reading data.
+
+ @param file The file containing the data or the file url of the data
+ */
+- (NSInputStream *)sdl_openInputStreamWithFile:(SDLFile *)file {
+    NSInputStream *inputStream = file.inputStream;
+    [inputStream setDelegate:self];
+    [inputStream scheduleInRunLoop:[NSRunLoop currentRunLoop]
+                                forMode:NSDefaultRunLoopMode];
+    [inputStream open];
+    return inputStream;
+}
 
-    // http://stackoverflow.com/a/503201 Make sure we get the exact number of packets we need
-    for (int i = 0; i < (((fileData.length - 1) / mtuSize) + 1); i++) {
-        SDLPutFile *putFile = [[SDLPutFile alloc] initWithFileName:file.name fileType:file.fileType persistentFile:file.isPersistent];
-        putFile.offset = @(currentOffset);
+/**
+ Returns the position in the file where to start reading data.
```

Comment:
I don't think this is correct, it's getting the length parameter for the `SDLPutFile` RPC, not the position to read from.